File: Sources/AnagramEngine/RackSolver.swift
import Foundation

public struct RackWord: Sendable, Identifiable {
    public let word: String
    public let score: Int      // Scrabble score (blanks count as 0)
    public let length: Int
    public let blanksUsed: Int
    public var id: String { word }
}

/// Finds every dictionary word that can be built from a subset of a letter rack,
/// optionally using blank tiles (wildcards) to fill missing letters.
public struct RackSolver: Sendable {
    let wordList: WordList
    public init(wordList: WordList) { self.wordList = wordList }

    /// - Parameters:
    ///   - rack: available letters; `?` or `*` characters count as blank tiles.
    ///   - minLength: ignore words shorter than this.
    /// - Returns: matching words sorted by score (desc), then length, then alpha.
    public func solve(rack: String, minLength: Int = 2) -> [RackWord] {
        let lower = rack.lowercased()
        let blanks = lower.filter { $0 == "?" || $0 == "*" }.count
        let rackLetters = LetterCount(lower)            // ignores ? and *
        let capacity = rackLetters.total + Int32(blanks)

        var out: [RackWord] = []
        out.reserveCapacity(2048)

        for w in wordList.words {
            let len = w.count
            if len < minLength { continue }
            if Int32(len) > capacity { continue }

            let wc = LetterCount(w)
            // How many letters does the rack lack? Those must be covered by blanks.
            var deficit = 0
            var blankScorePenalty = 0
            var ok = true
            for i in 0..<26 {
                let need = wc.counts[i] - rackLetters.counts[i]
                if need > 0 {
                    deficit += Int(need)
                    if deficit > blanks { ok = false; break }
                    // Letters covered by blanks score 0.
                    let ch = Character(UnicodeScalar(UInt8(97 + i)))
                    blankScorePenalty += Int(need) * Scrabble.score(String(ch))
                }
            }
            guard ok else { continue }

            let score = Scrabble.score(w) - blankScorePenalty
            out.append(RackWord(word: w, score: score, length: len, blanksUsed: deficit))
        }

        out.sort {
            if $0.score != $1.score { return $0.score > $1.score }
            if $0.length != $1.length { return $0.length > $1.length }
            return $0.word < $1.word
        }
        return out
    }
}

File: app/src/engine/rackSolver.ts
// Ported from Sources/AnagramEngine/RackSolver.swift
import { LetterCount } from "./letterCount";
import { WordList } from "./wordList";
import { scrabbleScore } from "./scrabble";

export interface RackWord {
  word: string;
  /** Scrabble score (blanks count as 0). */
  score: number;
  length: number;
  blanksUsed: number;
}

/**
 * Finds every dictionary word that can be built from a subset of a letter rack,
 * optionally using blank tiles (wildcards) to fill missing letters.
 */
export class RackSolver {
  private readonly wordList: WordList;

  constructor(wordList: WordList) {
    this.wordList = wordList;
  }

  /**
   * @param rack available letters; `?` or `*` characters count as blank tiles.
   * @param minLength ignore words shorter than this.
   * @returns matching words sorted by score (desc), then length, then alpha.
   */
  solve(rack: string, minLength = 2): RackWord[] {
    const lower = rack.toLowerCase();
    let blanks = 0;
    for (const ch of lower) if (ch === "?" || ch === "*") blanks++;
    const rackLetters = new LetterCount(lower); // ignores ? and *
    const capacity = rackLetters.total + blanks;

    const out: RackWord[] = [];
    for (const w of this.wordList.words) {
      const len = w.length;
      if (len < minLength) continue;
      if (len > capacity) continue;

      const wc = new LetterCount(w);
      // How many letters does the rack lack? Those must be covered by blanks.
      let deficit = 0;
      let blankScorePenalty = 0;
      let ok = true;
      for (let i = 0; i < 26; i++) {
        const need = wc.counts[i]! - rackLetters.counts[i]!;
        if (need > 0) {
          deficit += need;
          if (deficit > blanks) {
            ok = false;
            break;
          }
          // Letters covered by blanks score 0.
          blankScorePenalty += need * scrabbleScore(String.fromCharCode(97 + i));
        }
      }
      if (!ok) continue;

      const score = scrabbleScore(w) - blankScorePenalty;
      out.push({ word: w, score, length: len, blanksUsed: deficit });
    }

    out.sort((a, b) => {
      if (a.score !== b.score) return b.score - a.score;
      if (a.length !== b.length) return b.length - a.length;
      return a.word < b.word ? -1 : 1;
    });
    return out;
  }
}
